### PR TITLE
Improving webpki verifier CRL support ergonomics

### PR DESCRIPTION
**Pull request by cpu**
### Description

This branch updates Rustls to use the rustls-webpki v0.102.0-alpha.6 crate. This version of webpki improves CRL ergonomics (see https://github.com/rustls/webpki/pull/203). 

Since the changes in this alpha are breaking, I've taken the update & resolved the breaking changes in one commit to keep everything building and testing cleanly.

#### use `with_status_policy builder` fn

The upstream crate added a more ergonomic interface we can use in place of having to keep around a mutable builder and doing our own matching.

#### avoid CRL dyn trait hurdles

The upstream crate made working with CRLs easier by replacing the `CertRevocationList` trait with an `enum` representation.

Notably this makes working with the `Vec<OwnedCertRevocationList>` that the webpki verifier builders and verifiers hold much easier: we no long have to do as many contortions to convert to a `&[&dyn CertRevocationList]`.

## TODO

* [x] Wait for https://github.com/rustls/webpki/pull/203 to be merged
* [x] Wait for rustls-webpki v0.102.0-alpha.6 to be published
* [x] Remove `Cargo.toml` patch

### CODE REVIEW

**Comment by djc** [comment]
Would it be Hard to stack #1547 on top of this one instead of the other way around?

**Comment by cpu** [comment]
> Would it be Hard to stack https://github.com/rustls/rustls/pull/1547 on top of this one instead of the other way around?

I don't think it would be too hard, but it does mean that it can't be merged until the webpki parts land and we cut a new alpha (neither very big blockers). Do you feel strongly? #1547 has a +1 from Ctz so it feels closer to being mergeable than this and the webpki PR it uses.

**Comment by djc** [comment]
It seems like the proper order to me just because it means there will be a bunch of churn from porting the code added in #1547 to the new webpki API. But if you prefer to merge #1547 first I can definitely live with that.

**Comment by cpu** [comment]
> It seems like the proper order to me just because it means there will be a bunch of churn from porting the code added in https://github.com/rustls/rustls/pull/1547 to the new webpki API

I think 264bae2173b7697bb85f5edcd8ac81cbb5c9586b is the sum of the churn and it seems pretty minimal to me.

> In that case I will probably skip doing another round on https://github.com/rustls/rustls/pull/1547 and check back in with the rebased version of this.

Ok, that sounds good to me. I'm happy to revisit parts from #1547 if you find anything outside the scope of the rebased content.

**Comment by djc** [comment]
> I think [264bae2](https://github.com/rustls/rustls/commit/264bae2173b7697bb85f5edcd8ac81cbb5c9586b) is the sum of the churn and it seems pretty minimal to me.

Okay, that doesn't seem too bad!

File: rustls/src/webpki/mod.rs
Comment (djc) [inline rustls/src/webpki/mod.rs:112]: Let's make the `CertRevocationList` lifetime explicit here? (And the one for `CertificateRevocationListDer`, I suppose.)

Comment (cpu) [reply rustls/src/webpki/mod.rs]: Fixed!

**Comment by cpu** [comment]
Rebased on `main`, addressed feedback, and updated the PR desc. Once https://github.com/rustls/webpki/pull/203 lands I can flip this out of draft status and should be able to resolve the few failed CI tasks.

**Comment by cpu** [comment]
Leaving a reminder here: it'd be nice to get a new Rustls alpha after this lands so I can rework the FFI side in https://github.com/rustls/rustls-ffi/pull/341.

**Comment by cpu** [comment]
@djc This is ready for a once-over now if you're interested.

File: rustls/src/webpki/client_verifier.rs
Comment (djc) [inline rustls/src/webpki/client_verifier.rs:301]: Add a note about why this is okay?

File: rustls/src/webpki/server_verifier.rs
Comment (djc) [inline rustls/src/webpki/server_verifier.rs:253]: Here too.

**Review by djc — APPROVED** [review]
Did you want to bump the rustls alpha version so you can continue moving downstream into -ffi?

File: rustls/src/webpki/client_verifier.rs
Comment (cpu) [reply rustls/src/webpki/client_verifier.rs:303]: Good call. I think I had something like that in an earlier pass but lost it along the way. Fixed up.

File: rustls/src/webpki/server_verifier.rs
Comment (cpu) [reply rustls/src/webpki/server_verifier.rs:255]: Fixed too.

**Comment by cpu** [comment]
@ctz This could use a :mag: from you when you have a chance. Thanks!

**Review by ctz — APPROVED** [review]

**Comment by cpu** [comment]
> cpu force-pushed the cpu-1541-crl-for-server-cert-ergonomics branch from c12ce0f to 3c58a24 

Rebased to resolve conflicts after #1553 landed first.

**Pull request merged**
